plasmapy/classes/species.py

```diff
@@ -0,0 +1,231 @@
+                                         dtype=float) * u.m
+        self.velocity_history = np.zeros((self.NT, *self.v.shape),
+                                         dtype=float) * (u.m / u.s)
+        self._B_interpolator = RegularGridInterpolator(
+            (self.plasma.x.si.value,
+             self.plasma.y.si.value,
+             self.plasma.z.si.value),
+            self.plasma.magnetic_field.T.si.value,
+            method="linear",
+            bounds_error=True)
+
+        self._E_interpolator = RegularGridInterpolator(
+            (self.plasma.x.si.value,
+             self.plasma.y.si.value,
+             self.plasma.z.si.value),
+            self.plasma.electric_field.T.si.value,
+            method="linear",
+            bounds_error=True)
+
+    def _interpolate_fields(self):
+        interpolated_b = self._B_interpolator(self.x.si.value) * u.T
+        interpolated_e = self._E_interpolator(self.x.si.value) * u.V / u.m
+        return interpolated_b, interpolated_e
+
+    @property
+    def kinetic_energy(self):
+        """
+        Calculates the kinetic energy for each macroparticle.
+
+        Returns
+        --------
+        `astropy.units.Quantity`
+            Array of kinetic energies, shape (n,).
+        """
+        return (self.v ** 2).sum(axis=-1) * self.eff_m / 2
+
+    @property
+    def kinetic_energy_history(self):
+        """
+        Calculates the kinetic energy history for each particle.
+
+        Returns
+        --------
+        `astropy.units.Quantity`
+            Array of kinetic energies, shape (nt, n).
+        """
+        return (self.velocity_history ** 2).sum(axis=-1) * self.eff_m / 2
+
+    def boris_push(self, init=False):
+        """
+        Implements the Boris algorithm for moving particles and updating their
+        velocities.
+
+        Arguments
+        ----------
+        init: bool (optional)
+            If `True`, does not change the particle positions and sets dt
+            to -dt/2.
+        """
+        dt = -self.dt / 2 if init else self.dt
+        b, e = self._interpolate_fields()
+
+        # add first half of electric impulse
+        vminus = self.v + self.eff_q * e / self.eff_m * dt * 0.5
+
+        # rotate to add magnetic field
+        t = -b * self.eff_q / self.eff_m * dt * 0.5
+        s = 2 * t / (1 + t * t)
+        vprime = vminus + np.cross(vminus.si.value, t) * u.m / u.s
+        vplus = vminus + np.cross(vprime.si.value, s) * u.m / u.s
+
+        # add second half of electric impulse
+        v_new = vplus + self.eff_q * e / self.eff_m * dt * 0.5
+
+        self.v = v_new
+        if not init:
+            self.x += self.v * dt
+
+    def run(self):
+        """
+        Runs a simulation instance.
+        """
+        self.boris_push(init=True)
+        for i in range(self.NT):
+            self.boris_push()
+            self.position_history[i] = self.x
+            self.velocity_history[i] = self.v
```

Comment:
for i=0 this will overwrite the initial values that you're defining before the loop, no? (Not familiar with the algorith myself, maybe that's fine and I'm just showing my ignorance.)